components/locale/src/subtags/region.rs

```diff
@@ -0,0 +1,104 @@
+use crate::parser::errors::ParserError;
+use std::str::FromStr;
+use tinystr::TinyStr4;
+
+/// Region subtag (examples: `"US"`, `"CN"`, `"AR"` etc.)
+///
+/// `Region` represents a Unicode base language code conformat to the
+/// [`unicode_region_id`] field of the Language and Locale Identifier.
+///
+/// # Example
+/// ```
+/// use icu_locale::subtags::Region;
+///
+/// let region: Region = "DE".parse()
+///     .expect("Failed to parse a region subtag.");
+/// ```
+///
+/// [`unicode_region_id`]: https://unicode.org/reports/tr35/#unicode_region_id
+#[derive(Debug, PartialEq, Eq, Clone, Hash, PartialOrd, Ord, Copy)]
+pub struct Region(TinyStr4);
+
+const REGION_ALPHA_LENGTH: usize = 2;
+const REGION_NUM_LENGTH: usize = 3;
+
+impl Region {
+    /// A constructor which takes a utf8 slice, parses it and
+    /// produces a well-formed `Region`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu_locale::subtags::Region;
+    ///
+    /// let region = Region::from_bytes(b"fr")
+    ///     .expect("Parsing failed.");
+    ///
+    /// assert_eq!(region, "FR");
+    /// ```
+    pub fn from_bytes(v: &[u8]) -> Result<Self, ParserError> {
+        match v.len() {
+            REGION_ALPHA_LENGTH => {
+                let s = TinyStr4::from_bytes(v).map_err(|_| ParserError::InvalidSubtag)?;
+                if !s.is_ascii_alphabetic() {
+                    return Err(ParserError::InvalidSubtag);
+                }
+                Ok(Self(s.to_ascii_uppercase()))
```

Comment:
Hmm, I don't think I understand this change.

- How would it work for `"419"` if in line 48 you reject if its not alphabetic?
- How would it work for `"usa"` if you accept 3 letter alphabetic codes?

In order to combine them I'd have to make an if that is `if (len == 3 && numeric) || (len == 2 && alphabetic)`, unless I am missing something?

Reply:
The optimization I made here is that I first check the length, if it's not 2 or 3, I immediately bail, which allows for cheap skipping in case this is not a region at all.
If it is 2, then I parse it and verify that it's alphabetic, and if it's 3, I parse it and verify that it is numeric.
This way I only "pay" with parsing into `TinyStr` if the length is right and I'm already assuming that if the characters in it are wrong, that it's an error scenario and I don't have to optimize for it.

If I were to combine those branches, I'd have to either eagerly parse before I check the length, or test for alpha/num on `u8` rather than `TinyStr` (which is much slower).

Reply:
I tested the alternative:

```rust
    pub fn from_bytes(v: &[u8]) -> Result<Self, ParserError> {
        let len = v.len();
        if (len == REGION_ALPHA_LENGTH && v.iter().all(|b| b.is_ascii_alphabetic())) ||
            (len == REGION_NUM_LENGTH && v.iter().all(|b| b.is_ascii_digit())) {
            let s = TinyStr4::from_bytes(v).map_err(|_| ParserError::InvalidSubtag)?;
            Ok(Self(s.to_ascii_uppercase()))
        } else {
            Err(ParserError::InvalidSubtag)
        }
    }
```

and got `region_subtag_parse` regression from `74ns` to `81ns` and `construct` from `387ns` to `420ns`, so I don't think it's worth it.

Reply:
I'm going to resolve it and we can get back to it if we come up with a better approach.